flake.nix

```diff
@@ -1,83 +1,164 @@
+              inputsFrom = [
+                base
+              ];
               packages = [
-                (rust-bin.selectLatestNightlyWith (toolchain: toolchain.default.override {
-                  extensions = [
-                    "clippy"
-                    "llvm-tools-preview"
-                    "rust-src"
-                    "rustfmt"
-                  ];
-                  targets = [
-                    "wasm32-unknown-unknown"
-                    "x86_64-unknown-linux-musl"
-                    "x86_64-unknown-none"
-                  ];
-                }))
+                (rust-bin.selectLatestNightlyWith (toolchain: rustToolchain))
                 cargo-deadlinks
                 cargo-binutils
                 cargo-deny
+                cargo-fuzz
                 cargo-nextest
                 cargo-udeps
                 protobuf
                 qemu_kvm
               ];
             };
+            # Minimal shell with only the dependencies needed to run the format and check-format
+            # steps.
             lint = with pkgs; mkShell {
               packages = [
-                hadolint
+                bazel-buildtools
                 cargo-deadlinks
+                clang-tools
+                hadolint
+                nixpkgs-fmt
+                nodePackages.markdownlint-cli
                 nodePackages.prettier
                 nodePackages.prettier-plugin-toml
-                nixpkgs-fmt
+                shellcheck
               ];
+              shellHook = ''
+                export NODE_PATH=${nodePackages.prettier-plugin-toml}/lib/node_modules:$NODE_PATH
+              '';
             };
-            bazel = with pkgs; mkShell {
+            # Minimal shell with only the dependencies needed to run the bazel steps.
+            bazelShell = with pkgs; mkShell {
               shellHook = ''
-                export ANDROID_BASE_DIR=$(dirname $(dirname $(which android)))
-                export ANDROID_HOME=$ANDROID_BASE_DIR/libexec/android-sdk
+                export ANDROID_HOME="${androidSdk}/libexec/android-sdk"
+                export GRADLE_OPTS="-Dorg.gradle.project.android.aapt2FromMavenOverride=${androidSdk}/libexec/android-sdk/build-tools/28.0.3/aapt2";
               '';
               packages = [
-                jdk11
-                bazel_6
-                (androidenv.composeAndroidPackages {
-                  includeNDK = false;
-                  platformVersions = [ "30" "32" ];
-                  buildToolsVersions = [ "30.0.0" ];
-                }).androidsdk
+                jdk11_headless
                 bazel_6
+                androidSdk
+                bazel-buildtools
+              ];
+            };
+            # Shell for building Oak Containers kernel and system image. This is not included in the 
+            # default shell because it is not needed as part of the CI.
+            containers = with pkgs; mkShell {
```

Comment:
@andrisaar I added this shell with the extra tools needed to build the containers kernel and system image, and I added what appear to be the relevant input packages.

I am trying to build the kernel with

```console
cd oak_containers_kernel
make
```

But I get the following error:

```console
<stdin>:1:10: fatal error: libelf.h: No such file or directory
```

and other similar ones.

What is missing? And should I add it to this shell, now that the Dockerfile is not the source of truth for these deps?

Reply:
I tried adding https://search.nixos.org/packages?channel=unstable&show=libelf&from=0&size=50&sort=relevance&type=packages&query=libelf to the list of input packages, and now it goes a bit further, but it fails with

```
read_symbols':
elf.c:447:22: error: implicit declaration of function 'gelf_getsymshndx'; did you mean 'gelf_getsym'? [-Werror=implicit-function-declaration]
  447 |                 if (!gelf_getsymshndx(symtab->data, shndx_data, i, &sym->sym,
      |                      ^~~~~~~~~~~~~~~~
      |                      
```

perhaps it's trying to use a different version than the expected one?

Reply:
@mariaschett these are the deps you were asking about, they are already included in this dev shell, though it seems there are some incompatibilities with libelf.

ref https://github.com/project-oak/oak/pull/4062#issuecomment-1607515449

Reply:
I wasn't asking about any dependencies, I had to add these dependencies to the Docker Image

https://github.com/project-oak/oak/pull/4062/files

to get `Make` to run.

Reply:
Although it now fails later with a mysterious

```
make: *** [Makefile:7: target/linux-6.1.33/arch/x86/boot/bzImage] Error 2
```

Reply:
2 is `ENOENT`, is that weird nix sandboxing doing something so that the file can't be found?

Reply:
I'll try a couple of things and report back. I don't think this needs to block this PR though, since currently (i.e. before #4062 ) the kernel makefile does not work in the Docker image anyways.

Reply:
Fixed now, it turns out it was missing the `perl` binary from the dev shell :roll_eyes: